gdeep/trainer/regularizer.py

```diff
@@ -0,0 +1,432 @@
+        return regularizer_arguments
+
+    @abstractmethod
+    def update_params(self, pre_processed_arguments: Any) -> Any:
+        """
+        Any processing needed in a forward pass of the network
+        Anything that doesn't require gradients can go here
+        """
+        return pre_processed_arguments
+
+    @abstractmethod
+    def regularization_penalty(
+        self, forward_passed_arguments: Any, model: torch.nn.Module
+    ) -> Tensor:
+        """
+        Computes the actual regression penalty.
+        This must be an auto-differentiable function of its inputs
+        """
+        return torch.sum(forward_passed_arguments) ** 2
+
+
+class TopologicalRegularizer(Regularizer):
+    """
+    A topological regularizer for a 2-class classifier
+    """
+
+    def preprocess(self, **kwargs):
+        """
+        Initializes the topological regularizer:
+        kwargs:
+            n_neighbors: sizes of the KNN neighbors used for computing the local homology
+            hom_dim: the homologies to be computed
+            expansion:  The size of the clique-expansion when computing homology
+            regularization_dataset: Tensor describing where the linear surrogate function is evaluated
+        """
+
+        defaultKwargs = {
+            "n_neighbors": (6, 20),
+            "hom_dim": (0, 1, 2, 3, 4, 5, 6),
+            "expansion": 6,
+        }
+        self.regularizer_arguments = {**defaultKwargs, **kwargs}
+        print("Preprocessing regularizer")
+        dummy_loader = create_dummy_loader(
+            self.regularizer_arguments["regularization_dataset"]
+        )
+        connections = local_homology_preprocess(
+            self.regularizer_arguments["regularization_dataset"],
+            self.regularizer_arguments["n_neighbors"],
+            self.regularizer_arguments["hom_dim"],
+            dim=1,
+        )
+        self.regularizer_arguments["coords"] = self.regularizer_arguments[
+            "regularization_dataset"
+        ]
+        self.regularizer_arguments["connections"] = connections
+        self.regularizer_arguments["dummy_loader"] = dummy_loader
+        return True
+
+    def update_params(self, regularizer_arguments):
+        """
+        Finds the critical vertices of the linear surrogate function
+        """
+        network = self.regularizer_arguments["model"]
+        coords = self.regularizer_arguments["coords"]
+        dummy_loader = self.regularizer_arguments["dummy_loader"]
+        connections = self.regularizer_arguments["connections"]
+        expansion = self.regularizer_arguments["expansion"]
+        cpoints1 = compute_critical_points(
+            network, coords, dummy_loader, connections, expansion, ind=0
+        )
+        cpoints2 = compute_critical_points(
+            network, coords, dummy_loader, connections, expansion, ind=1
+        )
+        cpoints = np.concatenate(
+            [cpoints1.detach().numpy(), cpoints2.detach().numpy()], 0
+        )
```

Comment:
What's the difference between `cpoints`, `cpoints1` and `cpoints2`?

Reply:
The cpoints1 and cpoints2 are (select) critical points of f and -f, respectively. Cpoints is then both of them

Reply:
Could you please adjust the names to reflect that.

Reply:
done